### PR TITLE
send retry number on both GET and POST

### PR DESCRIPTION
**Pull request by Sarkar**
@aaustin @qinweigong @shortstuffsushi 

not particularly proud of the final originalUrl = url assignment, but without a decently sized re-write it's the cleanest way :/

### CODE REVIEW

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 1]
```diff
@@ -132,33 +140,41 @@ public ServerResponse make_restful_get(String url, String tag, int timeout, bool
...
 		try {
-			if (!addCommonParams(body)) {
+			if (!addCommonParams(body, retryNumber)) {
 				return new ServerResponse(tag, NO_BRANCH_KEY_STATUS);
 			}
 			
 			if (log) {
 				PrefHelper.Debug("BranchSDK", "posting to " + url);
-				PrefHelper.Debug("BranchSDK", "Post value = " + body.toString());
+				PrefHelper.Debug("BranchSDK", "Post value = " + body.toString(4));
```
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: just FYI, the 4 allows for prettier indenting to the console

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 2]
```diff
@@ -33,6 +33,7 @@
 
 	private static final String SDK_VERSION = "1.5.0";
 	private static final int DEFAULT_TIMEOUT = 3000;
+	private static int retryCount = 0;
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Should this be here?

Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: That's a relic of testing, good catch!

On Wednesday, April 15, 2015, Graham Mueller <notifications@github.com>
wrote:

> In Branch-SDK/src/io/branch/referral/RemoteInterface.java
> <https://github.com/BranchMetrics/Branch-Android-SDK/pull/61#discussion_r28473047>
> :
>
> > @@ -33,6 +33,7 @@
> >
> >  	private static final String SDK_VERSION = "1.5.0";
> >  	private static final int DEFAULT_TIMEOUT = 3000;
> > +	private static int retryCount = 0;
>
> Should this be here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BranchMetrics/Branch-Android-SDK/pull/61/files#r28473047>
> .
>


-- 

*Sahil Verma*

*Branch Metrics *

w branch.io | c 510-386-1750 | s sahilroyverma

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 3]
```diff
@@ -132,33 +139,41 @@ public ServerResponse make_restful_get(String url, String tag, int timeout, bool
...
+		return make_restful_post(body, url, tag, timeout, 0, true, linkData);
 	}
 
 	public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log) {
-		return make_restful_post(body, url, tag, timeout, log, null);
+		return make_restful_post(body, url, tag, timeout, 0, log, null);
 	}
 	
-	public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log, BranchLinkData linkData) {
+	public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout,
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: I think we had discussed this on the other PR, but should the method w/ the actual retryCount be public?

Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: that's a good point. it goes for both `make_restful_get` and this base post method, too. im totally fine w/ changing

**Comment by shortstuffsushi** [comment]
Regarding the `originalUrl` thing, I think it might be cleaner if you reversed it -- something like:

```
doGet(String baseUrl, ...) {
    String requestUrl = baseUrl;
    baseUrl += retry;
    baseUrl += ...

    // Make request

    if (resp.statusLine >= 500) {
       doGet(baseUrl)
    }
}
```

**Comment by shortstuffsushi** [comment]
Not modifying incoming params like that is kind of a pattern as well. If you can avoid reassign a variable that you received, probably for the best.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 4]
```diff
@@ -104,18 +105,24 @@ private boolean addCommonParams(JSONObject post) {
...
-			if (log) PrefHelper.Debug("BranchSDK", "getting " + url);
-		    HttpGet request = new HttpGet(url);
+		try {
+			if (log) PrefHelper.Debug("BranchSDK", "getting " + modifiedUrl);
+		    HttpGet request = new HttpGet(modifiedUrl);
 		    HttpResponse response = getGenericHttpClient(timeout).execute(request);
+			if (response.getStatusLine().getStatusCode() >= 500 &&
+				retryNumber < prefHelper_.getRetryCount()) {
+				retryNumber++;
+				make_restful_get(baseUrl, tag, timeout, retryNumber, log);
```
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: When the recursive calls unwind, they leave no change on the response object, isn't it?

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 5]
```diff
@@ -132,33 +139,41 @@ public ServerResponse make_restful_get(String url, String tag, int timeout, bool
...
+				PrefHelper.Debug("BranchSDK", "Post value = " + body.toString(4));
 			}
 		    HttpPost request = new HttpPost(url);
 		    request.setEntity(new ByteArrayEntity(body.toString().getBytes("UTF8")));
 		    request.setHeader("Content-type", "application/json");
 		    HttpResponse response = getGenericHttpClient(timeout).execute(request);
+            if (response.getStatusLine().getStatusCode() >= 500
+                    && retryNumber < prefHelper_.getRetryCount()) {
+                retryNumber++;
+                make_restful_post(body, url, tag, timeout, retryNumber, log, linkData);
```
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: same question as above.

**Comment by shortstuffsushi** [comment]
:+1:

**Comment by aaustin** [comment]
@Sarkar Two things:
- [ ] Since you're handling the retry in the RemoteInterface.java now, you'll also need to retrieve the timeout interval and Thread.Sleep for the allotted time before retrying
- [ ] Can you please clean up the other retry code in Branch.java. Currently, this will retry retryCount * retryCount times because it's still doing retries in the main file.

**Comment by Sarkar** [comment]
@aaustin removed the other retry code from Branch.java. I might need help with Thread.Sleep since I can't find an example of it. Is it just delaying execution for the set time for `Thread.Sleep`

**Pull request merged**